### PR TITLE
v0.2.3 client prediction

### PR DESCRIPTION
**Pull request by alexanderclarktx**
netcode
- server now populates action buffer by _predicting_ future actions for skelly
  - it will need a refactor (to be generic for any entity, configurable per action) but functionally it works
- client immediately copies future action buffer from server

graphics
- ball is now a sprite, not a text emoji

### CODE REVIEW

File: core/src/contrib/systems/net/WsServerSystem.ts [hunk 1]
```diff
@@ -3,22 +3,33 @@ import { System, TickData, World } from "@piggo-gg/core";
...
+
     frames.forEach((tick) => {
       const actionsAtTick = world.actionBuffer.atTick(tick);
       if (actionsAtTick && Object.keys(actionsAtTick).length) {
         actions[tick] = actionsAtTick;
+        Object.keys(actionsAtTick).forEach((playerId) => {
+          if (playerId.startsWith("skelly")) {
+            latestActionForEntity[playerId] = [tick, actionsAtTick[playerId]];
+          }
+        });
```
Comment (alexanderclarktx) [inline core/src/contrib/systems/net/WsServerSystem.ts:32]: this client prediction implementation needs refactored

whether an action should be "predicted" forward should be configurable. and it should be applied to all controlled entities generically

File: core/src/contrib/systems/ui/InputSystem.ts [hunk 2]
```diff
@@ -143,7 +141,7 @@ export const InputSystem: SystemBuilder<"InputSystem"> = ({
           const controllerInput = controller.controllerMap[input];
           if (controllerInput != null) {
             if (actions.actionMap[controllerInput]) {
-              world.actionBuffer.push(world.tick, controlledEntity.id, controllerInput);
+              world.actionBuffer.push(world.tick + 1, controlledEntity.id, controllerInput);
```
Comment (alexanderclarktx) [inline core/src/contrib/systems/ui/InputSystem.ts:144]: fixed 1 frame input delay (marginally improves networked entity positions abrupt start/stop)

File: Dockerfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-from oven/bun:1.0-slim
+from oven/bun:1.0.30-slim
```
Comment (alexanderclarktx) [inline Dockerfile:1]: bug in bun's newest docker image broke the build (perhaps related to this [issue](https://github.com/oven-sh/bun/issues/9428))

**Comment by DavFeng1** [comment]
8=============D

**Pull request merged**